File: src/main/java/Parser/Printer.java
package Parser;

import Lexer.Token;

public class Printer implements Expression.Visitor<String>{
    public String print(Expression exp) {
        return exp.accept(this);
    }

    @Override
    public String visitBinExp(Expression.BinExp exp) {
        return printNested(exp.operator, exp.left, exp.right);
    }

    @Override
    public String visitUnExp(Expression.UnExp exp){
        return printNested(exp.operator, exp.right);
    }

    @Override
    public String visitPrimeExp(Expression.PrimeExp exp) {
        return exp.name;
    }

    private String printNested(Token operator, Expression... exps) {
        StringBuilder string = new StringBuilder();
        string.append("(").append(operator.name).append(" (");

        for (Expression exp : exps) {
            string.append(exp.accept(this)).append(" ");
        }


        string.append("))");

        return string.toString();
    }
}

File: src/main/kotlin/parser/Printer.kt
package parser

import java.lang.StringBuilder
import lexer.Token
import parser.expressions.*
import parser.statements.*

class Printer : ExpVisitor<String>, StmtVisitor<String?> {

    fun print(stmt: Statement?): String {
        return stmt?.accept(this) ?: "null"
    }

    override fun visitBinExp(exp: BinExp): String {
        return printNested(exp.operator, exp.left, exp.right)
    }

    override fun visitUnExp(exp: UnExp): String {
        return printNested(exp.operator, exp.right)
    }

    override fun visitPrimeExp(exp: PrimeExp): String {
        return exp.name
    }

    private fun printNested(operator: Token?, vararg exps: Expression): String {
        val string = StringBuilder()
        string.append("(").append(operator!!.name).append(" (")
        for (exp in exps) {
            string.append(exp.accept(this)).append(" ")
        }
        string.append("))")
        return string.toString()
    }

    override fun visitExpStatement(stmt: ExpStatement): String {
        return stmt.exp.accept(this)
    }

    override fun visitReturnStatement(stmt: ReturnStatement): String {
        return "return ${stmt.exp.accept(this)}"
    }

    override fun visitJumpStatement(stmt: JumpStatement): String {
        return "jump to ${stmt.name}"
    }

    override fun visitIfStatement(stmt: IfStatement): String {
        return "if ${stmt.exp.accept(this)} ${stmt.firstJumpLabel.name} else ${stmt.firstJumpLabel.name}"
    }

    /*override fun visitVarStatement(stmt: VarStatement): String {
        return stmt.name
    }*/

    override fun visitAssignStatement(stmt: AssignStatement): String {
        return "${stmt.nameOfVariable} = ${stmt.exp.accept(this)}"
    }
}
